### PR TITLE
Pressure Recovery Ratio Feature: inter-foil gap position

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The inter-foil gap is the region of highest physics complexity in tandem configurations. The fore-foil wake mixes with the aft-foil leading-edge stagnation, creating a complex pressure recovery. A **normalized recovery ratio** r = (x - x_fore_TE) / (x_aft_LE - x_fore_TE), clamped to [0,1], explicitly tells the model where each node is in the inter-foil gap recovery stage. Zero for single-foil samples.

This targets p_tan specifically — the inter-foil pressure recovery is the dominant source of tandem-specific error. Only ~30 LoC.

### Changes to `cfd_tandemfoil/train.py`

1. **Add flag:**
   ```python
   parser.add_argument('--pressure_recovery_feature', action='store_true')
   ```

2. **Compute the feature:**
   ```python
   if args.pressure_recovery_feature:
       # For tandem samples:
       # fore_TE_x = max x-coordinate of boundary_id=6 (fore foil surface)
       # aft_LE_x = min x-coordinate of boundary_id=7 (aft foil surface)
       fore_te_x = x[boundary_id == 6, 0].max()  # x-coord of fore TE
       aft_le_x = x[boundary_id == 7, 0].min()   # x-coord of aft LE
       gap_length = aft_le_x - fore_te_x + 1e-6
       
       # Recovery ratio for ALL mesh nodes
       recovery_r = ((x[:, 0] - fore_te_x) / gap_length).clamp(0, 1)
       
       # Zero for single-foil samples (gap=0)
       if not is_tandem:
           recovery_r = torch.zeros_like(recovery_r)
       
       # Append as 1 additional input channel
   ```

3. **Update input_dim** by +1 to account for the new channel.

4. **Run 2 seeds:**
   ```bash
   CUDA_VISIBLE_DEVICES=0 python train.py \
     --agent askeladd --wandb_name "askeladd/recovery-ratio-s42" --wandb_group pressure-recovery-feature \
     --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
     --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
     --pressure_first --pressure_deep --residual_prediction --surface_refine \
     --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
     --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
     --pressure_recovery_feature

   CUDA_VISIBLE_DEVICES=1 python train.py \
     --agent askeladd --wandb_name "askeladd/recovery-ratio-s73" --wandb_group pressure-recovery-feature \
     --seed 73 [same flags]
   ```

## Baseline
| Metric | 2-seed avg | Target |
|--------|-----------|--------|
| **p_in** | **11.709** | < 11.71 |
| **p_oodc** | **7.544** | < 7.54 |
| **p_tan** | **27.402** | < 27.40 |
| p_re | 6.481 | < 6.48 |

W&B baseline: h6fqcry4 (s42), cuhoscp9 (s73)
Reproduce: `cd cfd_tandemfoil && python train.py --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 --slice_num 96 --cosine_T_max 150 --pcgrad_3way --pressure_first --pressure_deep --residual_prediction --surface_refine --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1`

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Metrics (2-seed average)

| Metric | Baseline | Seed 42 (`ex9llku7`) | Seed 73 (`sstip7ni`) | 2-seed avg | Δ vs baseline |
|--------|----------|----------------------|----------------------|------------|---------------|
| **p_in** | **11.709** | 12.2 | 12.6 | **12.4** | +5.9% ❌ |
| **p_oodc** | **7.544** | 7.7 | 7.6 | **7.65** | +1.4% ❌ |
| **p_tan** | **27.402** | 27.3 | 27.5 | **27.4** | flat |
| **p_re** | **6.481** | 6.7 | 6.4 | **6.55** | +1.1% ❌ |

### Commands

```bash
CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent askeladd --wandb_name "askeladd/recovery-ratio-s42" --wandb_group pressure-recovery-feature \
  --seed 42 --asinh_pressure --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --slice_num 96 --cosine_T_max 150 --pcgrad_3way \
  --pressure_first --pressure_deep --residual_prediction --surface_refine \
  --te_coord_frame --wake_deficit_feature --re_stratified_sampling --n_layers 3 \
  --cp_panel --cp_panel_tandem_only --cp_panel_scale 0.1 \
  --pressure_recovery_feature

# Same for seed 73
```

### W&B Run IDs
- Seed 42: `ex9llku7` (best epoch 159)
- Seed 73: `sstip7ni` (best epoch 160)

### Peak VRAM
~72 GB (no significant increase from baseline)

### Analysis

**The hypothesis did not pan out.** The pressure recovery ratio feature was designed to specifically target p_tan by providing the model with explicit positional information in the inter-foil gap. Results show:

- **p_tan (target metric): flat** — no improvement. The model already captures inter-foil gap geometry from existing features (raw coordinates, TE-relative coordinates, panel Cp).
- **p_in regressed +5.9%** — the extra input dimension slightly hurt interpolation generalization, likely because the recovery ratio adds a feature that's non-zero for tandem samples but zero for single-foil, creating a distribution shift the model must learn around.
- **p_oodc and p_re also slightly regressed** — consistent with the extra feature adding noise rather than signal.

The core issue: the model's existing feature set (24-dim input including TE-relative coordinates + panel Cp) already provides sufficient geometric context for the inter-foil region. The recovery ratio is a linearly derivable transformation of features the model already has (raw x-coordinate, fore-TE x, aft-LE x), so it doesn't add new information — it just adds redundancy that slightly degrades other metrics.

### Suggested follow-ups

1. **Learned inter-foil attention**: Instead of a handcrafted feature, use a cross-attention mechanism between fore-wake and aft-LE surface nodes to learn the inter-foil interaction directly.
2. **Tandem-specific loss weighting**: Weight tandem surface nodes in the gap region more heavily in the loss rather than adding features.
3. **Boundary-layer thickness feature**: A physics feature that captures the actual flow state (momentum thickness) rather than just geometric position might carry more information.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2354 because ADVISOR: Closing — p_in +5.9% regression with flat p_tan shows this feature isn't helping. The pressure recovery ratio (distance-along-surface / gap-width) is linearly derivable from the existing coordinate and geometry features the model already has access to. Unlike wake_deficit or cp_panel which encode physics the model can't easily learn from raw coordinates, this ratio adds redundant information. Thank you for testing this, askeladd..
